bahmni_api_feed/models/order_save_service.py

```diff
@@ -330,80 +330,162 @@ def _get_order_quantity(self, order, default_quantity_value, product_default_uom
 
     @api.model
     def _get_order_line_uom(self, order_line, product_default_uom):
-        
+
         uom_ids = self.env['syncable.units.mapping'].search([('name', '=', order_line['quantityUnits'])])
         if(uom_ids):
             uom_id = uom_ids.ids[0]
             uom_obj = self.env['syncable.units.mapping'].browse(uom_id)
             if uom_obj.unit_of_measure.id == product_default_uom.id:
                 return uom_obj.unit_of_measure.id
-        
+
         _logger.info("%s uom expected %s, but found %s"\
                 %(order_line['productName'],order_line['quantityUnits'],product_default_uom.name))
         return product_default_uom.id
+    @api.model
+    def get_available_batch_details(self, product_id, sale_order):
+        context = self._context.copy() or {}
+        sale_order = self.env['sale.order'].browse(sale_order)
+        context['location_id'] = sale_order.location_id and sale_order.location_id.id or False
+        context['search_in_child'] = True
+        shop_location_id = sale_order.shop_id.location_id.id if sale_order.shop_id.location_id.id else self.order_id.shop_id.location_id.id
+        stock_quant_lot = self.env['stock.quant'].search([
+            ('product_id','=', product_id.id if type(product_id) != list else product_id[0]),
+            ('location_id', '=', shop_location_id),
+            ('quantity', '>' , 0)
+        ])
+        already_used_batch_ids = [line.lot_id.id for line in sale_order.order_line if line.lot_id]
+        available_batches = {}
+        total_quantity = 0
+        for prodlot in stock_quant_lot:
+            if prodlot.lot_id.id not in already_used_batch_ids:
+                if prodlot.lot_id.expiration_date and prodlot.quantity > 0:
+                    date_length = len(str(prodlot.lot_id.expiration_date))
+                    formatted_ts = datetime.strptime(str(prodlot.lot_id.expiration_date), "%Y-%m-%d %H:%M:%S.%f").strftime("%Y-%m-%d %H:%M:%S") if date_length > 20 else datetime.strptime(str(prodlot.lot_id.expiration_date), "%Y-%m-%d %H:%M:%S").strftime("%Y-%m-%d %H:%M:%S")
+                    if formatted_ts and datetime.strptime(str(formatted_ts), DTF) > datetime.today():
+                        available_batches[prodlot.lot_id.id] = {'expiration_date': prodlot.lot_id.expiration_date, 'stock_forecast': prodlot.quantity}
```

Comment:
the reason for constructing a new object from stock.quant information in the form of the available_batches is to  Filter out already used batches to avoid duplication and Checking and formatting expiration dates.So that this object would only have necessary details for sale order.